### PR TITLE
BulletproofTxManager

### PR DESCRIPTION
**Pull request by samsondav**
This PR merges in a replacement transaction manager called BulletproofTxManager.

See design doc here: https://www.notion.so/chainlink/BulletproofTxManager-Architecture-Overview-9dc62450cd7a443ba9e7dceffa1a8d6b

This is a complete replacement for the legacy TxManager and doesn't use any of the original code. It features the following improvements:

1. Safe to run concurrently across multiple nodes (horizontally scalable)
2. Clear concurrency model (no race conditions)
3. Error handling is all done in one place and is clearer and easier to reason about
4. Clear separation of responsibilities and more comprehensible code
5. Guarantees around nonce tracking that **ensure** that there can never be gaps in the nonce sequence
6. Decoupling of `ethtx` and `txmanager` makes `ethtx` much smaller and simpler
7. Properly supports multiple private keys (node operators can now specify `fromAddress` in ethtx to pin a job to a particular address)
8. It is no longer required to ask for receipts on every new block, resulting in far fewer requests to the ethereum node (by an order of magnitude or more)
9. Deeper re-org protection (on the order of hundreds of blocks is possible)
10. Predictable behaviour in the event that an external account uses the nonce
11. Clear guarantees and design ensure transactions will ALWAYS be put on the longest canonical chain within well defined constraints.
12. Any new code can now easily send transactions simply by writing to the `eth_txes` table and letting BulletproofTxManager do the rest.

A new setting `ETH_FINALITY_DEPTH` is exposed which allows a node operator to specify how many blocks deep they want re-org protection for. This defaults to 50.

BulletproofTxManager is enabled by setting `ENABLE_BULLETPROOF_TX_MANAGER=true`. It is not recommended to switch this off and on, since there is a potential failure case, documented in the [release plan](https://www.notion.so/chainlink/Release-Plan-85b3891f8b674695a63545f5ff715499).

An escape hatch is provided with a new client command `setnextnonce`. This allows a user to manually set the next nonce in the sequence for a given account. It should never be necessary and is included "just in case".

### CODE REVIEW

File: core/adapters/eth_tx.go [hunk 1]
```diff
@@ -25,12 +26,13 @@ const (
 // EthTx holds the Address to send the result to and the FunctionSelector
 // to execute.
 type EthTx struct {
-	Address          common.Address       `json:"address"`
+	ToAddress        common.Address       `json:"address"`
+	FromAddress      *common.Address      `json:"fromAddress"`
 	FunctionSelector eth.FunctionSelector `json:"functionSelector"`
 	DataPrefix       hexutil.Bytes        `json:"dataPrefix"`
 	DataFormat       string               `json:"format"`
 	GasPrice         *utils.Big           `json:"gasPrice" gorm:"type:numeric"`
```
Comment (j16r) [inline core/adapters/eth_tx.go:34]: Does this need to be null? Pointer struct members actually make memory usage go up, so I wouldn't use them unless you really need null and it needs to be distinct from zero.

File: core/adapters/adapter.go [hunk 2]
```diff
@@ -87,7 +87,8 @@ func For(task models.TaskSpec, config orm.ConfigReader, orm *orm.ORM) (*Pipeline
 	if ba != nil { // task is for native adapter
 		err = unmarshalParams(task.Params, ba)
 	} else { // task is for external adapter
-		bt, err := orm.FindBridge(task.Type)
+		var bt models.BridgeType
```
Comment (j16r) [inline core/adapters/adapter.go:90]: What is the purpose of this change?

File: core/adapters/eth_tx.go [hunk 3]
```diff
@@ -42,6 +44,85 @@ func (e *EthTx) TaskType() models.TaskType {
...
+	} else {
+		return e.legacyPerform(input, store)
+	}
+}
+
+// TODO(sam): Move away from resuming this on every new head and do something more sensible
+func (e *EthTx) perform(input models.RunInput, store *strpkg.Store) models.RunOutput {
+	trtx, err := store.FindEthTaskRunTxByTaskRunID(input.TaskRunID().UUID())
+	if err != nil {
+		logger.Error(err)
```
Comment (j16r) [inline core/adapters/eth_tx.go:58]: Can we add some context here `logger.Error(err)` tends to be very challenging to use.

File: core/adapters/eth_tx.go [hunk 4]
```diff
@@ -42,6 +44,85 @@ func (e *EthTx) TaskType() models.TaskType {
...
+
+	var gasLimit uint64
+	if e.GasLimit == nil {
+		gasLimit = store.Config.EthGasLimitDefault()
+	} else {
+		gasLimit = *e.GasLimit
+	}
+
+	if err := store.IdempotentInsertEthTaskRunTx(taskRunID, fromAddress, toAddress, encodedPayload, gasLimit); err != nil {
+		logger.Error(err)
```
Comment (j16r) [inline core/adapters/eth_tx.go:116]: I'm noticing a pattern of handling errors with logging at multiple levels of the stack, I suspect when this code fails you'll get the same error being logged multiple times.

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 5]
```diff
@@ -0,0 +1,449 @@
...
+		if isVirginTransaction {
+			// If we get this error, and transaction has never been processed
+			// before, it is extremely likely that an external wallet messed with
+			// our nonce and we have no choice but to send it again.
+			//
+			// On the off chance an external wallet did not do this and we
+			// somehow got a network double-send or something, this fails by
+			// sending the transction twice (one will revert) which is a safe
+			// fail case.
+			return eb.handleExternalWalletUsedNonce(&etx, attempt)
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_broadcaster.go:265]: What if node operator restores from backup?

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 6]
```diff
@@ -0,0 +1,543 @@
...
+// ProcessHead takes all required transactions for the confirmer on a new head
+func (ec *ethConfirmer) ProcessHead(head models.Head) error {
+	return withAdvisoryLock(ec.store, ethConfirmerAdvisoryLockClassID, 0, func() error {
+		return ec.processHead(head)
+	})
+}
+
+// NOTE: This SHOULD NOT be run concurrently or it could behave badly
+func (ec *ethConfirmer) processHead(head models.Head) error {
+	ec.SetBroadcastBeforeBlockNum(head.Number)
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_confirmer.go:80]: unchecked error

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 7]
```diff
@@ -0,0 +1,543 @@
...
+		return err
+	}
+
+	etxs, err := FindEthTxsRequiringNewAttempt(ec.store.GetRawDB(), blockHeight, int64(ec.config.EthGasBumpThreshold()))
+	if err != nil {
+		return err
+	}
+	for _, etx := range etxs {
+		fmt.Println("attempts", etx.EthTxAttempts)
+		fmt.Println("old attempt", etx.EthTxAttempts[0].ID)
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_confirmer.go:195]: remove these

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 8]
```diff
@@ -0,0 +1,543 @@
...
+		return errors.New("attempt state must be in_progress")
+	}
+	return ec.store.GetRawDB().Save(attempt).Error
+}
+
+func (ec *ethConfirmer) handleInProgressAttempt(etx models.EthTx, a models.EthTxAttempt, blockHeight int64, isVirginAttempt bool) error {
+	if a.State != models.EthTxAttemptInProgress {
+		return fmt.Errorf("invariant violation: expected eth_tx_attempt %v to be in_progress, it was %s", a.ID, a.State)
+	}
+	fmt.Println("new attempt", a.ID)
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_confirmer.go:279]: remove this

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 9]
```diff
@@ -0,0 +1,543 @@
...
+	if a.ID == 0 {
+		return errors.New("expected attempt to have an id")
+	}
+	err := db.Exec(`DELETE FROM eth_tx_attempts WHERE id = ?`, a.ID).Error
+	logger.ErrorIf(err)
+	return err
+}
+
+// IsFinalized determines whether the transaction has an attempt that was
+// confirmed long enough ago that we consider it to be "final". Note that this
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_confirmer.go:415]: attempted*

File: core/store/orm/schema.go [hunk 10]
```diff
@@ -25,16 +25,18 @@ type ConfigSchema struct {
 	DefaultHTTPTimeout              models.Duration `env:"DEFAULT_HTTP_TIMEOUT" default:"15s"`
 	Dev                             bool            `env:"CHAINLINK_DEV" default:"false"`
 	EnableExperimentalAdapters      bool            `env:"ENABLE_EXPERIMENTAL_ADAPTERS" default:"false"`
+	EnableBulletproofTxManager      bool            `env:"ENABLE_BULLETPROOF_TX_MANAGER" default:"false"`
```
Comment (se3000) [inline core/store/orm/schema.go:28]: Is this intentionally different from MIN_INCOMING_CONFIRMATIONS?

File: core/adapters/eth_tx.go [hunk 11]
```diff
@@ -25,12 +26,13 @@ const (
...
 type EthTx struct {
-	Address          common.Address       `json:"address"`
+	ToAddress        common.Address       `json:"address"`
+	FromAddress      *common.Address      `json:"fromAddress"`
 	FunctionSelector eth.FunctionSelector `json:"functionSelector"`
 	DataPrefix       hexutil.Bytes        `json:"dataPrefix"`
 	DataFormat       string               `json:"format"`
 	GasPrice         *utils.Big           `json:"gasPrice" gorm:"type:numeric"`
-	GasLimit         uint64               `json:"gasLimit"`
+	GasLimit         *uint64              `json:"gasLimit"`
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/adapters/eth_tx.go:35]: Because it's optional?

File: core/adapters/eth_tx.go (hunk 1 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:34]: I figured pointers were used when something could be omitted. It's unambiguous to check for `nil`.

You can legitimately have 0-priced transactions, they sometimes crop up on Ropsten.

File: core/adapters/adapter.go (hunk 2 above)
Comment (samsondav) [reply core/adapters/adapter.go:90]: `go vet` yelled at me about it

File: core/adapters/eth_tx.go (hunk 3 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:58]: Yeah, I thought about it but figured the stacktrace would be enough.

File: core/adapters/eth_tx.go (hunk 4 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:116]: You are correct we do log at multiple levels, and will get multiple errors.

I added the `logger.Error` at the lowest level since it helps debugging to see exactly where the error is coming from.

Any of these errors would be pretty serious so I don't see an issue in logging it multiple times... but if that is a problem, how would you prefer to solve the problem of logging being specific to the point of occurrence?

**Comment by samsondav** [comment]
@se3000 `rebroadcasttransactions` command should be entirely redundant now since the new tx manager is reliable.

File: core/store/orm/schema.go (hunk 10 above)
Comment (samsondav) [reply core/store/orm/schema.go:28]: Which one.. do you mean `EthFinalityDepth`?

Yes, it is intentionally different.

File: core/adapters/eth_tx.go (hunk 3 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:58]: Perhaps you could give an example of what you have in mind?

File: core/adapters/eth_tx.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (j16r) [reply core/adapters/eth_tx.go:35]: Sure, but could 0 represent optional and then we don't need nil? If null represents optional, what does 0 mean?

File: core/adapters/eth_tx.go (hunk 4 above)
Comment (j16r) [reply core/adapters/eth_tx.go:121]: We can and should be using Cheney's `errors` package which gets us stack traces, he also talks about how errors should only be logged once: https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 12]
```diff
@@ -0,0 +1,449 @@
...
+		} else {
+			var wg sync.WaitGroup
+
+			// It is safe to process separate keys concurrently
+			// NOTE: This design will block one key if another takes a really long time to execute
+			for _, key := range keys {
+				wg.Add(1)
+				go func(k models.Key) {
+					if err := eb.ProcessUnstartedEthTxs(k); err != nil {
+						// NOTE: retries if this function errors are unbounded,
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_broadcaster.go:132]: Need to reword this.

File: core/adapters/eth_tx.go (hunk 3 above)
Comment (j16r) [reply core/adapters/eth_tx.go:58]: Generally I like to have something like `logger.Error("error finding task by task run id", err)`

File: core/adapters/eth_tx.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (samsondav) [reply core/adapters/eth_tx.go:35]: I guess it's pretty certain that you don't actually want to send a transaction with a gas limit of 0. I can change this back.

File: core/adapters/eth_tx.go (hunk 3 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:58]: Yup, I can do this

File: core/adapters/eth_tx.go (hunk 4 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:120]: Ah, this looks like a much better solution. I'll use this errors package and wrap the errors at the bottom level, then we get a stacktrace at the top when they're logged out.

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 13]
```diff
@@ -0,0 +1,449 @@
...
+		//
+		// It is however possible an external wallet can have messed with the
+		// account during restart.
+		//
+		// In this case, the onus is on the node operator since this is
+		// explicitly unsupported.
+		//
+		// If it turns out to have been an external wallet, this transaction
+		// will be retried on every new head (and will fail) until we reach out
+		// block depth target and then abandoned.
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_broadcaster.go:279]: Make it more clear that this does nothing

File: core/services/bulletprooftxmanager/eth_broadcaster.go [hunk 14]
```diff
@@ -0,0 +1,449 @@
...
+	} else if sendError.IsTerminallyUnderpriced() {
+		bumpedGasPrice := BumpGas(eb.config, attempt.GasPrice.ToInt())
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_broadcaster.go:281]: Consider pulling this out

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 15]
```diff
@@ -0,0 +1,547 @@
...
+func (ec *ethConfirmer) CheckForReceipts() error {
+	// NOTE: If this becomes a performance bottleneck due to eth node requests,
+	// it may be possible to use goroutines here to speed it up
+	unconfirmedEtxs, err := ec.findUnconfirmedEthTxs()
+	if err != nil {
+		return err
+	}
+	for _, etx := range unconfirmedEtxs {
+		for _, attempt := range etx.EthTxAttempts {
+			receipt, err := ec.fetchReceipt(attempt.Hash)
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_confirmer.go:112]: Can we optimise since previous nonces will always be confirmed?

File: core/services/head_tracker.go [hunk 16]
```diff
@@ -132,7 +124,7 @@ func (ht *HeadTracker) Save(h models.Head) error {
 	if err != nil {
 		return err
```
Comment (samsondav) [inline core/services/head_tracker.go:125]: what if we take too long to process the head? will it block?

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 17]
```diff
@@ -0,0 +1,547 @@
...
+		// never mark the transaction as confirmed.
+		//
+		// The simplest and safest thing to do in that case is to keep
+		// requesting receipts on each head until one of two things happens:
+		//
+		// 1. We get a receipt for one of our transactions
+		// 2. We hit such a high block height that whichever transaction is
+		// there won't get re-org'd out so it doesn't matter anyway (for
+		// gapless nonce purposes).
+		if ec.IsFinalized(etx, blockHeight) {
```
Comment (samsondav) [inline core/services/bulletprooftxmanager/eth_confirmer.go:341]: Can we just finalize when we get a later nonce?

File: core/adapters/eth_tx.go [hunk 18]
```diff
@@ -42,6 +44,93 @@ func (e *EthTx) TaskType() models.TaskType {
...
+		return e.checkForConfirmation(*trtx, input, store)
+	}
+	return e.insertEthTx(input, store)
+}
+
+func (e *EthTx) checkForConfirmation(trtx models.EthTaskRunTx, input models.RunInput, store *store.Store) models.RunOutput {
+	switch trtx.EthTx.State {
+	case models.EthTxConfirmed:
+		receipt := models.EthReceipt{}
+		if err := store.GetRawDB().
```
Comment (se3000) [inline core/adapters/eth_tx.go:72]: Why do we need to skip the advisory lock here?

File: core/adapters/eth_tx.go [hunk 19]
```diff
@@ -42,6 +44,93 @@ func (e *EthTx) TaskType() models.TaskType {
...
+		return models.NewRunOutputComplete(output)
+	case models.EthTxFatalError:
+		return models.NewRunOutputError(trtx.EthTx.GetError())
+	default:
+		return models.NewRunOutputPendingOutgoingConfirmationsWithData(input.Data())
+	}
+}
+
+func (e *EthTx) insertEthTx(input models.RunInput, store *store.Store) models.RunOutput {
+	value, err := getTxData(e, input)
```
Comment (se3000) [inline core/adapters/eth_tx.go:94]:
```suggestion
	txData, err := getTxData(e, input)
```

"Value" is another field in an Etheruem transaction, the naming is a bit confusing as is.

File: core/adapters/eth_tx.go [hunk 20]
```diff
@@ -42,6 +44,93 @@ func (e *EthTx) TaskType() models.TaskType {
...
+		fromAddress = *e.FromAddress
+	} else {
+		fromAddress, err = store.GetDefaultAddress()
+		if err != nil {
+			err = errors.Wrap(err, "insertEthTx failed to GetDefaultAddress")
+			logger.Error(err)
+			return models.NewRunOutputError(err)
+		}
+	}
+	encodedPayload := utils.ConcatBytes(e.FunctionSelector.Bytes(), e.DataPrefix, value)
```
Comment (se3000) [inline core/adapters/eth_tx.go:113]:
```suggestion
	encodedPayload := utils.ConcatBytes(e.FunctionSelector.Bytes(), e.DataPrefix, txData)
```

File: core/adapters/eth_tx.go (hunk 18 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:72]: I'd flip it round... why would we need an advisory lock here?

File: core/adapters/eth_tx.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (spooktheducks) [reply core/adapters/eth_tx.go:35]: You can do `,omitempty` in the json struct tag and it will omit the value when it's 0.

File: core/adapters/eth_tx.go (hunk 1 above)
Comment (spooktheducks) [reply core/adapters/eth_tx.go:34]: Re: `nil`-able fields --
Benefit: increase in clarity
Cost: node panics (or at least ends up in a weird state, if it's a sub-goroutine that panics) when a single task run has a bad input

Personally, I don't care about memory in cases like these, but I think the Go philosophy of preferring zero values over nil-able values is probably healthier for the software, as long as we have good error reporting.

**Comment by spooktheducks** [comment]
> @se3000 no, it does not work with `rebroadcasttransactions`.
> 
> I am not really sure if there is any useful reason to make it work with that command either.
> 
> The new tx manager operates in such a way that it basically enforces that all transactions get broadcast eventually anyway... so I don't know what purpose that command would serve.

I agree with this approach, but also want to know -- is there some reason we may want to be able to force rebroadcasting?  For example, if the ClintEastwoodTxManager implicitly imposes some restrictions that create a time delay on rebroadcasting, and we want to get around that during an emergency scenario.

File: core/adapters/eth_tx.go (hunk 18 above)
Comment (spooktheducks) [reply core/adapters/eth_tx.go:72]: Yeah, I agree that we should start thinking of things in terms of "why lock" rather than "why not lock" if we want to get to our goal of horizontal scalability.  But at the same time, can you answer your own question for our (and posterity's) edification?

File: core/adapters/eth_tx.go (hunk 1 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:34]: @j16r @spooktheducks this reasoning makes sense to me, will convert to use zero values instead of nils

File: core/adapters/eth_tx.go (hunk 18 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:72]: OK, I'll try to explain. There is no need to take the Big Global Lock (tm) for this particular query because there aren't any concurrency issues.

It's a read only (no writes) and it doesn't matter when it happens. All state updates for `eth_txes` et al are atomic enforced via transaction so there is no danger of reading inconsistent data. If we read before the receipt gets written, we remain unconfirmed. If a receipt then gets written, we will confirm on the next head.

**Comment by samsondav** [comment]
@spooktheducks it's a fair point, I will adjust the rebroadcaster to work with the new tx manager.

File: core/adapters/eth_tx.go (hunk 18 above)
Comment (samsondav) [reply core/adapters/eth_tx.go:72]: As I wrote this out I suddenly realised we _can_ be in an inconsistent state, because we emit two queries non-transactionally:

1. Fetch ethtx
2. Switch on state
3. Fetch receipt

The state can be modified between 1 and 3 if the reorg detector marks it as unconfirmed.

Then in step 3 we would look for a receipt for an unconfirmed transaction. This probably wouldn't be the end of the world (we don't ever delete receipts) and the tx would simply be marked confirmed, but it isn't perfect behaviour. Will fix this.

Good catch @se3000.

**Comment by samsondav** [comment]
@se3000 @spooktheducks I started work on the rebroadcaster, it's a little more involved than the old one because it must be able to run in parallel with a running node and there are some concurrency considerations.

How do you feel about the following:

1. Document a rollback procedure
2. Roll this out to a couple of node operators on testnet as-is so we can start getting some airtime with it early
3. In the meantime I'll work on the rebroadcast command and ship that as a followup

?

**Comment by samsondav** [comment]
Check all usage of TxManager

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 21]
```diff
@@ -0,0 +1,543 @@
...
+	if err := ec.BumpGasWhereNecessary(head.Number); err != nil {
+		return err
+	}
+
+	return ec.EnsureConfirmedTransactionsInLongestChain(head)
+}
+
+func (ec *ethConfirmer) SetBroadcastBeforeBlockNum(blockNum int64) error {
+	return ec.store.GetRawDB().Exec(
+		`UPDATE eth_tx_attempts SET broadcast_before_block_num = ? WHERE broadcast_before_block_num IS NULL AND state = 'broadcast'`,
```
Comment (se3000) [inline core/services/bulletprooftxmanager/eth_confirmer.go:95]: Use prepared statement parameter here.

File: core/cmd/local_client.go [hunk 22]
```diff
@@ -439,6 +439,32 @@ func (cli *Client) DeleteUser(c *clipkg.Context) (err error) {
 	return err
 }
 
+// SetNextNonce manually updates the keys.next_nonce field for the given key with the given nonce value
+func (cli *Client) SetNextNonce(c *clipkg.Context) error {
+	addressHex := c.String("address")
+	nextNonce := c.Uint64("nextNonce")
+
+	logger.SetLogger(cli.Config.CreateProductionLogger())
+	db, err := gorm.Open("postgres", cli.Config.DatabaseURL())
```
[the remark this answers is not in the document]
Comment (topliceanu) [reply core/cmd/local_client.go:448]: The local client connects to the database :o

File: core/internal/cltest/mocks.go [hunk 23]
```diff
@@ -69,6 +69,20 @@ func MockEthOnStore(t testing.TB, s *store.Store, flags ...string) *EthMock {
 	return mock
 }
 
+// SimpleGethWrapper offers an easy way to mock the eth client
+type SimpleGethWrapper struct {
+	c eth.GethClient
+}
+
+func NewSimpleGethWrapper(c eth.GethClient) *SimpleGethWrapper {
```
Comment (topliceanu) [inline core/internal/cltest/mocks.go:77]: Is this used anywhere?

File: core/services/bulletprooftxmanager/errors.go [hunk 24]
```diff
@@ -0,0 +1,128 @@
...
+// Fatal errors mean that no matter how many times the send is retried, no node
+// will ever accept it
+func (s *sendError) Fatal() bool {
+	return s != nil && s.fatal
+}
+
+// Parity errors
+var (
+	// Non-fatal
+	parTooCheapToReplace    = regexp.MustCompile("^Transaction gas price .+is too low. There is another transaction with same nonce in the queue")
```
Comment (topliceanu) [inline core/services/bulletprooftxmanager/errors.go:35]: typo here!

Comment (samsondav) [reply core/services/bulletprooftxmanager/errors.go:35]: Where?

File: core/internal/cltest/mocks.go (hunk 23 above)
Comment (samsondav) [reply core/internal/cltest/mocks.go:77]: In tests

File: core/adapters/eth_tx.go [hunk 25]
```diff
@@ -42,6 +46,113 @@ func (e *EthTx) TaskType() models.TaskType {
...
+}
+
+func (e *EthTx) checkForConfirmation(trtx models.EthTaskRunTx, input models.RunInput, store *store.Store) models.RunOutput {
+	switch trtx.EthTx.State {
+	case models.EthTxConfirmed:
+		return checkEthTxForReceipt(trtx.EthTx.ID, input, store)
+	case models.EthTxFatalError:
+		return models.NewRunOutputError(trtx.EthTx.GetError())
+	default:
+		return models.NewRunOutputPendingOutgoingConfirmationsWithData(input.Data())
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/adapters/eth_tx.go:81]: This is a legacy hangover from the way the runmanager works... It's a bit wild but it calls into `Perform` on every new head. I opted to keep it that way for now and will fix later.

It could totally get here in `unstarted` or `in_progress` - in either case it's still ok to be in `PendingOutgoingConfirmations`. We just have to wait for it to get to either `fatal_error` or `unconfirmed/confirmed`

[the remark this answers is not in the document]
Comment (samsondav) [reply core/adapters/eth_tx.go:81]: The run manager is consuming this output, to know whether its safe to go onto the next task yet.

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 26]
```diff
@@ -0,0 +1,543 @@
...
+	if a.State != models.EthTxAttemptInProgress {
+		return errors.New("deleteInProgressAttempt: expected attempt state to be in_progress")
+	}
+	if a.ID == 0 {
+		return errors.New("deleteInProgressAttempt: expected attempt to have an id")
+	}
+	return errors.Wrap(db.Exec(`DELETE FROM eth_tx_attempts WHERE id = ?`, a.ID).Error, "deleteInProgressAttempt failed")
+}
+
+// IsFinalized determines whether the transaction has an attempt that was
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/eth_confirmer.go:411]: How about `IsSafeToAbandon`?

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 27]
```diff
@@ -0,0 +1,543 @@
...
+func (ec *ethConfirmer) Connect(*models.Head) error {
+	return nil
+}
+
+func (ec *ethConfirmer) Disconnect() {
+	// pass
+}
+
+func (ec *ethConfirmer) OnNewLongestChain(head models.Head) {
+	if ec.config.EnableBulletproofTxManager() {
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/eth_confirmer.go:62]: Hmm... no good reason I suppose it was more of a "belts and braces" approach.

We can simply not register it instead.

File: core/services/bulletprooftxmanager/bulletprooftxmanager.go [hunk 28]
```diff
@@ -0,0 +1,154 @@
...
+	return signedTx.Hash(), rlp.Bytes(), nil
+
+}
+
+// send broadcasts the transaction to the ethereum network, writes any relevant
+// data onto the attempt and returns an error (or nil) depending on the status
+func sendTransaction(gethClientWrapper strpkg.GethClientWrapper, a models.EthTxAttempt) *sendError {
+	signedTx, err := a.GetSignedTx()
+	if err != nil {
+		return FatalSendError(err)
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/bulletprooftxmanager.go:74]: Yeah probably since if we fail at this point it is unlikely that any number of retries will ever succeed.

File: core/cmd/local_client.go (hunk 22 above)
[the remark this answers is not in the document]
Comment (samsondav) [reply core/cmd/local_client.go:448]: It could be run while the node is running, although it is probably safer if the node is not running.

Nonce does not exist in memory the source of truth is always the database.

I will update to use `orm.DialectPostgresWithoutLock` :+1:

File: core/store/orm/orm.go [hunk 29]
```diff
@@ -239,25 +244,40 @@ func (orm *ORM) AllSyncEvents(cb func(*models.SyncEvent) error) error {
 	})
 }
 
+// NOTE: Copied verbatim from gorm master
+// Transaction start a transaction as a block,
+// return error will rollback, otherwise to commit.
+func (orm *ORM) Transaction(fc func(tx *gorm.DB) error) (err error) {
```
Comment (se3000) [inline core/store/orm/orm.go:250]: What's the benefit of copying this in?

File: core/adapters/eth_tx.go (hunk 25 above)
[the remark this answers is not in the document]
Comment (topliceanu) [reply core/adapters/eth_tx.go:81]: Got it!

File: core/services/bulletprooftxmanager/eth_confirmer.go [hunk 30]
```diff
@@ -0,0 +1,543 @@
...
+		return errors.New("expected eth_tx state to be confirmed")
+	}
+	return errors.Wrap(db.Exec(`UPDATE eth_txes SET state = 'unconfirmed' WHERE id = ?`, etx.ID).Error, "unconfirmEthTx failed")
+}
+
+func unbroadcastAttempt(db *gorm.DB, a models.EthTxAttempt) error {
+	if a.State != models.EthTxAttemptBroadcast {
+		return errors.New("expected eth_tx_attempt to be broadcast")
+	}
+	return errors.Wrap(db.Exec(`UPDATE eth_tx_attempts SET broadcast_before_block_num = NULL, state = 'in_progress' WHERE id = ?`, a.ID).Error, "unbroadcastAttempt failed")
```
[the remark this answers is not in the document]
Comment (samsondav) [reply core/services/bulletprooftxmanager/eth_confirmer.go:547]: On balance I'd prefer to keep the status hardcoded. Using prepared statements here would make it less readable for no real gain.

File: core/services/bulletprooftxmanager/eth_confirmer.go (hunk 21 above)
Comment (samsondav) [reply core/services/bulletprooftxmanager/eth_confirmer.go:95]: See previous comment about prepared statements for the state.

File: core/services/bulletprooftxmanager/errors.go (hunk 24 above)
Comment (topliceanu) [reply core/services/bulletprooftxmanager/errors.go:35]: I misread that line, it's a regexp, not a typo.
